osquery/tables/system/windows/groups.cpp

```diff
@@ -0,0 +1,139 @@
+
+// Defined in /osquery/core/windows/process_ops.cpp, but not declared in a
+// header:
+std::string psidToString(PSID sid);
+
+// Get the relative identifier (RID) from a security identifier (SID):
+unsigned long getRidFromSid(PSID sidPtr) {
+  BYTE* countPtr = GetSidSubAuthorityCount(sidPtr);
+  unsigned long indexOfRid = static_cast<unsigned long>(*countPtr - 1);
+  unsigned long* ridPtr = GetSidSubAuthority(sidPtr, indexOfRid);
+  return *ridPtr;
+}
+
+namespace tables {
+
+std::unique_ptr<BYTE[]> GetSid(LPCWSTR accountName) {
+  // Validate the input parameters.
+  if (accountName == nullptr) {
+    LOG(INFO) << "GetSid(): no account name provided.";
+    return nullptr;
+  }
+
+  // Call LookupAccountNameW() once to retrieve the necessary buffer sizes for
+  // the SID (in bytes) and the domain name (in TCHARS):
+  unsigned long sidBufferSize = 0;
+  unsigned long domainNameSize = 0;
+  auto eSidType = SidTypeUnknown;
+  LookupAccountNameW(nullptr,
+                     accountName,
+                     nullptr,
+                     &sidBufferSize,
+                     nullptr,
+                     &domainNameSize,
+                     &eSidType);
+
+  // Allocate sufficient buffers for the (binary data) SID and the (wide string)
+  // domain name:
+  auto sidBuffer = std::make_unique<BYTE[]>(sidBufferSize);
+  std::vector<wchar_t> domainName(domainNameSize);
+
+  // Call LookupAccountNameW() a second time to actually obtain the SID for the
+  // given account name:
+  auto ret = LookupAccountNameW(nullptr,
+                                accountName,
+                                sidBuffer.get(),
+                                &sidBufferSize,
+                                domainName.data(),
+                                &domainNameSize,
+                                &eSidType);
+  if (ret == 0) {
+    LOG(INFO) << "Failed to LookupAccountNameW(): " << accountName;
+  } else if (IsValidSid(sidBuffer.get()) == FALSE) {
+    LOG(INFO) << "The SID for " << accountName << " is invalid.";
+  }
+
+  return sidBuffer; // Implicit move operation. Caller "owns" returned pointer.
+}
+
+void processLocalGroups(QueryData& results) {
+  unsigned long groupInfoLevel = 1;
+  unsigned long numGroupsRead = 0;
+  unsigned long totalGroups = 0;
+  unsigned long resumeHandle = 0;
+  unsigned long ret = 0;
+  LOCALGROUP_INFO_1* lginfo = nullptr;
+
+  std::unique_ptr<BYTE[]> sidSmartPtr = nullptr;
+  PSID sidPtr = nullptr;
+
+  do {
+    ret = NetLocalGroupEnum(nullptr,
+                            groupInfoLevel,
+                            (LPBYTE*)&lginfo,
+                            MAX_PREFERRED_LENGTH,
+                            &numGroupsRead,
+                            &totalGroups,
+                            nullptr);
+
+    if ((ret == NERR_Success || ret == ERROR_MORE_DATA) && lginfo != nullptr) {
+      for (size_t i = 0; i < numGroupsRead; i++) {
+        Row r;
+        sidSmartPtr = GetSid(lginfo[i].lgrpi1_name);

```

Comment:
This could be a `nullptr`, you should `continue` if it is.

Reply:
Ok I think I see what you mean. I'll push a new commit that addresses this.

Reply:
I'm confused, most likely due to lack of experience with these APIs. But I want to double-check.

You're setting this `sidPtr` to `nullptr`. Then passing it to `getRidFromSid(sidPtr)` even if it's a `nullptr`. That function will call https://msdn.microsoft.com/en-us/library/windows/desktop/aa446658(v=vs.85).aspx which expects a valid SID.

Reply:
You are right, this error handling is incomplete and I will have to revisit it. 

Earlier, I may have been unsure whether it was possible to retrieve a list of group names from `NetLocalGroupEnum` for which any one of the entries would return an invalid SID from `LookupAccountNameW`. I think I was intending to display a group name with a blank for SID. But, I can't find any documented case where that should happen, so maybe the better thing to do is, if `LookupAccountNameW` fails, don't add the row to the table at all and just move on. Maybe log a warning.